### PR TITLE
Abstract http client and support xhr and fetch backends

### PR DESCRIPTION
**Pull request by tokland**
Required by https://github.com/EyeSeeTea/metadata-synchronization/issues/655

### CODE REVIEW

File: src/api/common.ts [hunk 1]
```diff
@@ -167,6 +166,4 @@ export interface D2ApiDefinitionBase {
     filter: Record<string, any>;
 }
 
-export type D2ApiResponse<Data> = CancelableResponse<Data>;
-
-export const D2ApiResponse = CancelableResponse;
+export { CancelableResponse as D2ApiResponse } from "../repositories/CancelableResponse";
```
Comment (SferaDev) [inline src/api/common.ts:169]: Simplified the renamed export of ``D2ApiResponse``

File: src/api/d2Api.ts [hunk 2]
```diff
@@ -41,7 +38,7 @@ export class D2ApiGeneric {
         return this.apiConnection.getMockAdapter();
     }
 
-    public request<T>(options: D2ApiRequest): CancelableResponse<T> {
+    public request<T>(options: D2ApiRequest): D2ApiResponse<T> {
```
Comment (SferaDev) [inline src/api/d2Api.ts:41]: Use ``D2ApiResponse`` inside ``api`` folder to make it consistent.

File: src/api/d2Api.ts [hunk 3]
```diff
@@ -142,3 +139,5 @@ export abstract class D2ApiVersioned<
         return new Email(this);
     }
 }
+
+export { D2ApiOptions };
```
Comment (SferaDev) [inline src/api/d2Api.ts:143]: Re-export ``D2ApiOptions``

File: src/react/useD2ApiData.ts [hunk 4]
```diff
@@ -1,8 +1,8 @@
 import axios, { Canceler } from "axios";
 import { useCallback, useEffect, useState } from "react";
-import { CancelableResponse } from "../repositories/CancelableResponse";
+import { D2ApiResponse } from "../api/common";
 
-export type D2ApiDataHookQuery<Data> = Pick<CancelableResponse<Data>, "cancel" | "getData">;
+export type D2ApiDataHookQuery<Data> = Pick<D2ApiResponse<Data>, "cancel" | "getData">;
```
Comment (SferaDev) [inline src/react/useD2ApiData.ts]: Use ``D2ApiResponse`` inside ``react`` folder to make it consistent.

File: src/repositories/CancelableResponse.ts [hunk 5]
```diff
@@ -54,3 +54,9 @@ export class CancelableResponse<Data> {
 const noop = () => {
     return;
 };
+
+export type Canceler = (message?: string) => void;
+
+export function isCancel(error: any) {
+    return Axios.isCancel(error) || (error && error.name === "AbortError");
+}
```
Comment (SferaDev) [inline src/repositories/CancelableResponse.ts:62]: Export ``isCancel`` method

File: src/repositories/CancelableResponse.ts [hunk 6]
```diff
@@ -54,3 +54,9 @@ export class CancelableResponse<Data> {
 const noop = () => {
     return;
 };
+
+export type Canceler = (message?: string) => void;
```
Comment (SferaDev) [inline src/repositories/CancelableResponse.ts:58]: Export ``Canceler`` type so that we do not rely on ``Axios`` anymore on the app side

**Review by SferaDev — APPROVED** [review]
Everything looks good to me!

Added some minor changes and commented them so you can review them

File: src/repositories/CancelableResponse.ts (hunk 5 above)
Comment (tokland) [reply src/repositories/CancelableResponse.ts:62]: Could go as `api.isCancel()` (which uses the repository), but that's also ok.

**Comment by SferaDev** [comment]
One last thing that just crossed my mind, we do not have a polyfill for node. So we should either:

- Throw an error if we detect ``node`` is the runner
- Add a polyfill version (just import it on the api file)

**Pull request merged**